### ISSUE TITLE
Running tests ?

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#2277]
It seems there's a `build-test` field to build the tests. Aren't we missing a `run-test` field ?

**Comment by seliopou** [comment]
Perhaps poorly named, but as I understand it, `build-test` should build and run the tests.

**Comment by dbuenzli** [comment]
There still seems to be no good story on that front in OPAMv2. 

The current scheme has the disavantage of often leading to duplicate build instructions in both `build-test:` and `build:` which is not DRY (this is particularly acute in `topkg`-based packages but also happens for `configure`-based packages). Besides it would also be good to allow to distinguish between building and running the tests (tant pis for those who run during build).

Here are suggestions:

**a.** Add a `build-test` or `test` (to be consistent with `test` deps) filter and variable in `build` so that packagers can specify all their build at a single place in the `build:` section. For example: 
```
  ["./configure" "--prefix=%{prefix}%" "--enable-tests" {test} ]
```
or 
```
  ["ocaml" "pkg/pkg.ml" "build" "--tests" "%{test}%" --pinned "%{pinned}%"]
```
**b.**  Provide a `run-test:` field with a list of commands to execute the test. 

The cli command story can be figured out later. But here are a few hints:
currently `-t` is equivalent of `--build-test` and executes `build-test:`.

The option `-t` and an equivalent `--run-test` should be kept for build and run (i.e. set `test` to true and execute both `build-test:` and `run-test:` whenever they exist).  

For `install/upgrade` it's unclear if there's a utility in distinguishing at the cli level between build and run. If not then `--build-test` should be deprecated. If there is then `--build-test` should simply build (i.e. set the `test` variable to `true` and invoke `build-test:` if it exists but not `run-test:`).

**Comment by AltGr** [comment]
Indeed, the recursive behaviour of the `-t` flag has been fixed, but this is still an open issue, and we should get a story for 2.0.

- replacing `build-test:` with conditions in `build:`: sounds fine to me, whatever is more convenient. I can define the variable and either leave `build-test:` as it is or deprecate it
- adding a `run-test:` field sounds fine to me, in practice this would actually be a rename of the `build-tests:` field...

**Comment by AltGr** [comment]
Note: `build-doc` should probably be removed and handled in a similar fashion (filters in `build:`; filters in `install:` would probably be more sensible than an `install-doc:` field)

**Comment by AltGr** [comment]
Hm, we have a naming consistency issue:
- in `depends:`, boolean vars `build`, `doc`, `test` are used
- in `build:`, variable `doc` is already defined to point to the package's documentation directory, so we would have to use `build-doc`, and consistently `build-test`, but that's not consistent with the above.

`doc` as a directory is already used throughout, while the `doc` depends flag doesn't seem to be used much yet, so we might rename the `depends:` flags... although having `doc` for the flag and `docdir` (and similarly `libdir`, etc.) would be less error-prone IMHO.

**Comment by dbuenzli** [comment]
Maybe the simplest is to have the variables as `build-doc`, `build-test` in `build:`.

**Comment by AltGr** [comment]
We now have `build-test` and `build-doc` variables, available both during `build:` and `install:`.
Missing:
- they are not properly documented
- the similarly named fields are still there

So we have two ways of doing the same thing, and it's still not clear when the tests are _run_.
We could remove the fields and add a rewrite path to put them into `build:` with the proper filter.

Nothing difficult, but I am not sure what will be most convenient.
Proposal:
- `build-test:` now means `run-test:`. Maybe rename the field. Instructions contained there can be split into the `build:` part (using filters) and the `run-test:` part later on, for new packages: it doesn't cause problems if the test building is in `run-test:`, it's just inconvenient.
- `build-doc` should be split between `build:` and `install:`, with the proper filters, and the field removed. That split is hard to do automatically though. Most packages seem to just build the doc in `build-doc:`, and have the installation handled by the general `install:`, maybe we can grep and fix the few others by hand.

**Comment by AltGr** [comment]
Closed by 4170148
